tests/Endpoint/PartitionTest.php

```diff
@@ -0,0 +1,403 @@
+            [$partition, 'buzz', 'fizz', 'https://fizz.buzz.bar'],
+            [$partition, 'pop', 'fizz', 'https://pop.fizz.bar'],
+            [$partition, 'us-east-1', 'iot', 'https://iot.us-east-1.bar'],
+        ];
+    }
+
+    /**
+     * @dataProvider partitionDefinitionProvider
+     *
+     * @param array $definition
+     */
+    public function testSupportsArrayAccess(array $definition)
+    {
+        $partition = new Partition($definition);
+        $this->assertSame($definition, $partition->toArray());
+        $partition['foo'] = 'bar';
+        $this->assertTrue(isset($partition['foo']));
+        $this->assertSame('bar', $partition['foo']);
+        unset($partition['foo']);
+        $this->assertFalse(isset($partition['foo']));
+    }
+
+    /**
+     * @dataProvider signatureVersionProvider
+     *
+     * @param Partition $partition
+     * @param $region
+     * @param $service
+     * @param $signatureVersion
+     */
+    public function testDeterminesSignatureVersion(
+        Partition $partition,
+        $region,
+        $service,
+        $signatureVersion
+    ) {
+        $resolved = $partition([
+            'scheme' => 'https',
+            'service' => $service,
+            'region' => $region,
+        ]);
+
+        $this->assertArrayHasKey('signatureVersion', $resolved);
+        $this->assertSame($signatureVersion, $resolved['signatureVersion']);
+    }
+
+    public function signatureVersionProvider()
+    {
+        $partition = new Partition([
+            'partition' => 'foo',
+            'dnsSuffix' => 'bar',
+            'defaults' => ['signatureVersions' => ['v4']],
+            'regions' => [],
+            'services' => [
+                'baz' => [
+                    'isRegionalized' => false,
+                    'partitionEndpoint' => 'foo-global',
+                    'endpoints' => [
+                        'foo-global' => ['signatureVersions' => ['s3v4']],
+                    ],
+                ],
+                'fizz' => [
+                    'defaults' => ['signatureVersions' => ['anonymous']],
+                    'endpoints' => [
+                        'buzz' => [],
+                        'pop' => [
+                            'signatureVersions' => ['s3v4'],
+                        ],
+                    ],
+                ],
+                'quux' => [
+                    'endpoints' => [
+                        'puff' => [],
+                    ],
+                ],
+            ],
+        ]);
+
+        return [
+            // partition endpoint setting
+            [$partition, 'us-east-1', 'baz', 's3v4'],
+            // service setting
+            [$partition, 'buzz', 'fizz', 'anonymous'],
+            // service endpoint setting
+            [$partition, 'pop', 'fizz', 's3v4'],
+            // no overrides
+            [$partition, 'puff', 'quux', 'v4'],
+            // unknown service
+            [$partition, 'us-east-1', 'iot', 'v4'],
```

Comment:
How is this one resolving? "us-east-1" doesn't match the region explicitly nor do I see a region regex that it would match? Is this a test for an unknown service and and unknown region (e.g., fallback to assuming the aws partition as a last resort)? If so, maybe you should add a test that is just for unknown service, and one that is just for unknown region.